rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,412 @@
+		Header:         *header,
+		ccc:            ccc,
+		coinbase:       coinbase,
+		state:          state,
+		gasPool:        new(core.GasPool).AddGas(header.GasLimit),
+	}
+}
+
+func (p *Pipeline) WithBeforeTxHook(beforeTxHook func()) *Pipeline {
+	p.beforeTxHook = beforeTxHook
+	return p
+}
+
+func (p *Pipeline) Start(deadline time.Time) error {
+	p.start = time.Now()
+	p.TxnQueue = make(chan *types.Transaction)
+	applyStageRespCh, incrementCh, err := p.traceAndApplyStage(p.TxnQueue)
+	if err != nil {
+		log.Error("Failed starting traceAndApplyStage", "err", err)
+		return err
+	}
+	p.applyStageRespCh = applyStageRespCh
+	p.ResultCh = p.cccStage(incrementCh, deadline)
+	return nil
+}
+
+func (p *Pipeline) TryPushTxns(txs types.OrderedTransactionSet, onFailingTxn func(txnIndex int, tx *types.Transaction, err error) bool) *Result {
+	for {
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+
+		result, err := p.TryPushTxn(tx)
+		if result != nil {
+			return result
+		}
+
+		switch {
+		case err == nil, errors.Is(err, core.ErrNonceTooLow):
+			txs.Shift()
+		default:
+			if errors.Is(err, ErrApplyStageDone) || onFailingTxn(p.txs.Len(), tx, err) {
+				close(p.TxnQueue)
+				p.TxnQueue = nil
+				return nil
+			}
+
+			if tx.IsL1MessageTx() {
+				txs.Shift()
+			} else {
+				txs.Pop()
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *Pipeline) TryPushTxn(tx *types.Transaction) (*Result, error) {
+	select {
+	case p.TxnQueue <- tx:
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+
+	select {
+	case err, valid := <-p.applyStageRespCh:
+		if !valid {
+			return nil, ErrApplyStageDone
+		}
+		return nil, err
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+}
+
+func (p *Pipeline) Kill() {
+	if p.TxnQueue != nil {
+		close(p.TxnQueue)
+	}
+
+	select {
+	case <-p.applyStageRespCh:
+		<-p.ResultCh
+	case <-p.ResultCh:
+		<-p.applyStageRespCh
+	}
+}
+
+type PendingBlockIncrement struct {
+	LastTrace      *types.BlockTrace
+	NextL1MsgIndex uint64
+
+	// accumulated state
+	Header        *types.Header
+	State         *state.StateDB
+	Txs           types.Transactions
+	Receipts      types.Receipts
+	CoalescedLogs []*types.Log
+}
+
+func (p *Pipeline) traceAndApplyStage(txsIn <-chan *types.Transaction) (chan error, chan *PendingBlockIncrement, error) {
+	p.state.StartPrefetcher("miner")
+	incrementCh := make(chan *PendingBlockIncrement)
+	resCh := make(chan error)
+	go func() {
+		defer func() {
+			close(incrementCh)
+			close(resCh)
+			p.state.StopPrefetcher()
+		}()
+
+		var tx *types.Transaction
+		for {
+			applyIdleTimer.Time(func() {
+				tx = <-txsIn
+			})
+			if tx == nil {
+				return
+			}
+
+			applyStart := time.Now()
+
+			// If we don't have enough gas for any further transactions then we're done
+			if p.gasPool.Gas() < params.TxGas {
+				return
+			}
+
+			// If we have collected enough transactions then we're done
+			// Originally we only limit l2txs count, but now strictly limit total txs number.
```

Comment:
> `// Originally we only limit l2txs count, but now strictly limit total txs number.`

Why?

Reply:
That is not a new comment, I copied it over from the original worker.